File: server/scripts/download_stock_data.py
import yfinance as yf
import sys
import os
from datetime import datetime, timedelta

def download_stock_data(ticker, days, output_path):
    try:
        end_date = datetime.now()
        start_date = end_date - timedelta(days=days)
        stock = yf.Ticker(ticker)
        df = stock.history(start=start_date, end=end_date)
        
        if df.empty:
            print(f"No data found for {ticker}")
            return False
            
        # Ensure directory exists
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        
        # Save to CSV
        df.to_csv(output_path)
        print(f"Downloaded {len(df)} records for {ticker}")
        return True
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return False

if __name__ == "__main__":
    if len(sys.argv) < 4:
        print("Usage: python download_stock_data.py <ticker> <days> <output_path>")
        sys.exit(1)
        
    ticker = sys.argv[1]
    days = int(sys.argv[2])
    output = sys.argv[3]
    success = download_stock_data(ticker, days, output)
    sys.exit(0 if success else 1)

File: server/scripts/download_stock_data.js
const yahooFinance = require('yahoo-finance2').default;
const fs = require('fs').promises;
const path = require('path');

async function downloadStockData(ticker, days, outputPath) {
  try {
    const endDate = new Date();
    const startDate = new Date();
    startDate.setDate(endDate.getDate() - days);

    const queryOptions = {
      period1: startDate.toISOString().split('T')[0],
      period2: endDate.toISOString().split('T')[0],
    };

    const result = await yahooFinance.historical(ticker, queryOptions);

    if (!result || result.length === 0) {
      console.log(`No data found for ${ticker}`);
      return false;
    }

    // Convert to CSV
    const headers = Object.keys(result[0]).join(',');
    const rows = result.map(row => Object.values(row).join(','));
    const csvContent = `${headers}\n${rows.join('\n')}`;

    // Ensure directory exists
    await fs.mkdir(path.dirname(outputPath), { recursive: true });

    // Save to CSV
    await fs.writeFile(outputPath, csvContent);
    console.log(`Downloaded ${result.length} records for ${ticker}`);
    return true;
  } catch (e) {
    console.error(`Error: ${e.message}`, e);
    return false;
  }
}

if (require.main === module) {
  if (process.argv.length < 5) {
    console.log("Usage: node download_stock_data.js <ticker> <days> <output_path>");
    process.exit(1);
  }

  const [, , ticker, days, outputPath] = process.argv;
  const daysInt = parseInt(days, 10);

  if (isNaN(daysInt)) {
    console.error("Error: <days> must be an integer.");
    process.exit(1);
  }

  downloadStockData(ticker, daysInt, outputPath).then(success => {
    process.exit(success ? 0 : 1);
  });
}
